pkg/converter/exporter.go

```diff
@@ -15,46 +15,232 @@
-	numElements := len(e.Objects)
+// ObjectsAsYAML converts cluster objects into single YAML file.
+func (e *Exporter) ObjectsAsSingleYAML() (string, error) {
 	var sb strings.Builder
-	for i, o := range e.Objects {
-		yaml, err := FromObject(o).ToYAML()
+	numComp := len(e.comps)
+	for j, c := range e.comps {
+		numObj := len(c.Spec.Objects)
+		for i, o := range c.Spec.Objects {
+			yaml, err := FromObject(o).ToYAML()
+			if err != nil {
+				return "", err
+			}
+			sb.Write(yaml)
+			if i < numObj-1 || j < numComp-1 {
+				// Join the objects into one document.
+				// Note: Each doc ends with a newline (from the ToYAML step), so we don't
+				// need to write an additional newline
+				sb.WriteString("---\n")
+			}
+		}
+	}
+	return sb.String(), nil
+}
+
+// ComponentSet creates a component set with the given name and version.
+func (e *Exporter) ComponentSet(setName, setVersion string) (*bundle.ComponentSet, error) {
+	if setName == "" {
+		return nil, errors.New("set name is required to export ComponentSet, but was empty")
+	}
+	if setVersion == "" {
+		return nil, errors.New("set version is required to export ComponentSet, but was empty")
+	}
+	newBun := &bundle.Bundle{
+		SetName:    setName,
+		Version:    setVersion,
+		Components: e.comps,
+	}
+	return newBun.ComponentSet(), nil
+}
+
+// ComponentsAsSingleYAML creates a single YAML file, delimited by '---'. If
+// a component set is defined, the component set is append to the end.
+func (e *Exporter) ComponentsAsSingleYAML(set *bundle.ComponentSet) (string, error) {
+	var sb strings.Builder
+	for i, c := range e.comps {
+		cyaml, err := FromObject(c).ToYAML()
 		if err != nil {
-			return "", err
+			return "", fmt.Errorf("while rendering component %v: %v", c.ComponentReference(), err)
 		}
-		sb.Write(yaml)
-		if i < numElements-1 {
-			// Join the objects into one document.
-			// Note: Each doc ends with a newline (from the ToYAML step), so we don't
-			// need to write an additional newline
+		sb.Write(cyaml)
+		if i < len(e.comps)-1 {
 			sb.WriteString("---\n")
 		}
 	}
+	if set != nil {
+		sb.WriteString("---\n")
+		cyaml, err := FromObject(set).ToYAML()
+		if err != nil {
+			return "", fmt.Errorf("while rendering component set: %v", err)
+		}
+		sb.Write(cyaml)
+	}
 	return sb.String(), nil
 }
+
+// ComponentsWithPathTemplates takes in a set of pathTemplates and an optional
+// set, and produces a map of paths to YAML files.
+//
+// pathTemplates are template strings and must be a templated filePath. The
+// optional values are
+//
+// - ComponentName: name of component
+// - Version: version of the component or component set
+// - BuildTag: build tag for the component
+// - SetName: name of the component set
+//
+// For example, here are same possible ways to craft pathTemplates
+//
+//   "{{.ComponentName}}/{{.Version}}/{{.ComponentName}}-{{.Version}}.yaml",
+//   "{{.ComponentName}}/{{.BuildTag}}/{{.ComponentName}}.yaml",
+//   "sets/{{.SetName}}-{{.Version}}.yaml",
```

Comment:
Non-deterministic paths is really going to complicate tracking these things safely.

Reply:
I'm not following -- how is it non-deterministic?

Reply:
s/non-determinisitic/non-stable/. Stability in export paths greatly simplifies the task of automatically tracking these things over time. Can you describe a scenario where this would be useful?

Reply:
Are you suggesting that this shouldn't be configurable? I thought about doing that

Reply:
No, just trying to understand the value of templating this vs requiring it to be a fully specified path.

Reply:
Can you give an example? The templating was so I could support variable tags that got used as part of the path. And also because I couldn't commit to a particular format.